### PR TITLE
feat(tls.Server) basic support (cert, key, rejectUnauthorized, requestCert, ca)

### PR DESCRIPTION
**Pull request by cirospaciari**
This depends on https://github.com/Jarred-Sumner/uSockets/pull/1
To run the tests, rebuild usockets with the PR source is needed.

Closes https://github.com/oven-sh/bun/issues/2504

### CODE REVIEW

**Comment by Jarred-Sumner** [comment]
I think uWS will need to also be updated to point to the new build of uSockets

**Comment by cirospaciari** [comment]
> I think uWS will need to also be updated to point to the new build of uSockets

Yeah i will do that, In another PR we should add a customization so uWS can use the key, cert in memory stuff too

File: src/bun.js/api/server.zig [hunk 1]
```diff
@@ -183,6 +249,34 @@ pub const ServerConfig = struct {
                     any = true;
                 }
             }
+
+            if (obj.getTruthy(global, "key")) |js_obj| {
+                if (js_obj.jsType().isArray()) {
```
Comment (Jarred-Sumner) [inline src/bun.js/api/server.zig:254]: can we throw an error in `else` here to say that it is not valid?

Comment (Jarred-Sumner) [reply src/bun.js/api/server.zig:254]: "Expected `key` to be an array of strings`

File: src/bun.js/api/server.zig [hunk 2]
```diff
@@ -183,6 +249,34 @@ pub const ServerConfig = struct {
...
+                    const count = js_obj.getLengthOfArray(global);
+                    if (count > 0) {
+                        var native_array = bun.default_allocator.alloc([*c]const u8, count) catch unreachable;
+                        result.key = native_array.ptr;
```
Comment (Jarred-Sumner) [inline src/bun.js/api/server.zig:258]: the array of pointers will need to be freed on error right

File: src/bun.js/api/server.zig [hunk 3]
```diff
@@ -183,6 +249,34 @@ pub const ServerConfig = struct {
...
+
+                        var i: u32 = 0;
+                        var valid_count: u32 = 0;
+
+                        while (i < count) : (i += 1) {
+                            var item = JSC.JSObject.getIndex(js_obj, global, i);
+
+                            var sliced = item.toSlice(global, bun.default_allocator);
+                            defer sliced.deinit();
```
Comment (Jarred-Sumner) [inline src/bun.js/api/server.zig:267]: `sliced` can be `const`

maybe we can do `toSliceZ`? and then we don't need to clone it potentially twice

File: src/bun.js/api/server.zig [hunk 4]
```diff
@@ -183,6 +249,34 @@ pub const ServerConfig = struct {
...
+                    const count = js_obj.getLengthOfArray(global);
+                    if (count > 0) {
+                        var native_array = bun.default_allocator.alloc([*c]const u8, count) catch unreachable;
+                        result.key = native_array.ptr;
+
+                        var i: u32 = 0;
+                        var valid_count: u32 = 0;
+
+                        while (i < count) : (i += 1) {
+                            var item = JSC.JSObject.getIndex(js_obj, global, i);
```
Comment (Jarred-Sumner) [inline src/bun.js/api/server.zig:264]:
```suggestion
                            const item = JSC.JSObject.getIndex(js_obj, global, i);
```

we should just add `getIndex` to `JSValue` I think. Then this becomes `js_obj.getIndex(global, i)`

File: src/bun.js/node-tls.exports.js [hunk 5]
```diff
@@ -78,31 +117,110 @@ const TLSSocket = (function (InternalTLSSocket) {
...
   },
 );
+
+class Server extends NetServer {
+  key;
+  cert;
+  ca;
+  passphrase;
+  secureOptions;
+  _rejectUnauthorized;
```
Comment (Jarred-Sumner) [inline src/bun.js/node-tls.exports.js:140]: Does Node expose `_requestCert` and `_rejectUnauthorized`? If not can we make this `#rejectUnauthorized` and `#requestCert`

File: src/bun.js/api/server.zig (hunk 1 above)
Comment (Jarred-Sumner) [reply src/bun.js/api/server.zig:254]: actually I think we should do

`JSC.Node.StringOrBuffer` for the element type in these

That lets us accept:
- string
- any typed array view
- arraybuffer

technically we should also support Bun.file() for these, but we can leave that for another day

**Comment by Jarred-Sumner** [comment]
need to rebase, couple comments but looks close to mergeable

**Comment by Jarred-Sumner** [comment]
The files changed is really hard to read. Can you push up a rebased version?

**Pull request closed without merging**